File: latticeai/core/agent/recovery.py
"""ROLLBACK and MEMORY — putting the workspace back, and keeping the lesson.

Recovery tries git first where the policy says the tool is git-rollbackable,
then the pre-write snapshot the executor captured, and otherwise reports
``mode="none"`` rather than claiming a recovery it did not perform. The memory
phase records what a *terminal* run taught: DONE runs record what worked,
FAILED and NEEDS_REVIEW runs record what went wrong, stored under the run's
actual terminal state and never a blanket "ok".
"""

from __future__ import annotations

import json
import logging
from typing import Any, Dict, List, Optional

from latticeai.core.agent_helpers import extract_action, filter_learnings
from latticeai.core.agent_state import AGENT_TERMINAL_STATES, AgentState

from ._contract import AgentCore as _Core
from .context import AgentRunContext


class _RecoveryMixin(_Core):
    """The ROLLBACK and MEMORY phases of :class:`SingleAgentRuntime`."""

    # ── ROLLBACK ─────────────────────────────────────────────────────
    def _snapshot_for(self, ctx: AgentRunContext, path: str) -> Optional[Dict[str, Any]]:
        for entry in ctx.rollback_log:
            if entry.get("path") == path:
                return entry
        return None

    def _rollback_one(self, ctx: AgentRunContext, path: str, gov: Dict[str, Any]) -> Dict[str, Any]:
        """Recover one path: git when governed and available, else the
        pre-write snapshot, else an honest ``mode="none"`` (review L7)."""
        d = self.deps
        if gov.get("rollback") == "git" and d.rollback_file is not None:
            try:
                result = dict(d.rollback_file(str(path)))
            except Exception as exc:  # noqa: BLE001
                result = {"path": path, "ok": False, "error": str(exc)}
            if result.get("ok"):
                result["mode"] = "git"
                return result
        snapshot = self._snapshot_for(ctx, str(path))
        if snapshot is not None and d.restore_snapshot is not None and not snapshot.get("too_large"):
            content = snapshot.get("content") if snapshot.get("existed") else None
            try:
                restored = dict(d.restore_snapshot(str(path), content))
            except Exception as exc:  # noqa: BLE001
                restored = {"path": path, "ok": False, "error": str(exc)}
            restored.setdefault("path", path)
            restored["mode"] = "snapshot"
            return restored
        return {
            "path": path, "ok": False, "mode": "none",
            "error": "no rollback available (git not applicable, no usable snapshot)",
        }

    def rollback(self, ctx: AgentRunContext, current_user: str) -> None:
        """ROLLBACK: recover written files (git → snapshot → none), then FAILED."""
        d = self.deps
        rolled: List[dict] = []
        seen_paths: set = set()
        for step in ctx.transcript:
            if step.get("state") != AgentState.EXECUTING.value:
                continue
            if not isinstance(step.get("result"), dict):
                continue
            gov = step.get("governance", {}) or {}
            path = step["result"].get("path") or (step.get("args") or {}).get("path", "")
            if not path or str(path) in seen_paths:
                continue
            if gov.get("rollback") != "git" and step.get("action") not in d.file_create_actions:
                continue
            seen_paths.add(str(path))
            rolled.append(self._rollback_one(ctx, str(path), gov))

        ctx.transcript.append({"state": AgentState.ROLLBACK.value, "rolled_back": rolled})
        ctx.trace.decision(
            "rollback", decision="rolled_back",
            attempted=len(rolled), recovered=sum(1 for r in rolled if r.get("ok")),
        )
        recovered = [f"{r['path']} ({r.get('mode')})" for r in rolled if r.get("ok")]
        ctx.final_message = (
            f"실행 실패로 롤백했습니다. 복구 파일: {recovered}"
            if recovered
            else "롤백을 시도했으나 복구할 파일이 없거나 git/스냅샷 복구 수단이 없습니다."
        )
        d.audit("agent_rollback", user_email=current_user, rolled_back=rolled)
        self._emit_step(ctx, "rollback", "rolled_back", recovered=len(recovered))
        # Rollback is a recovery from a failed verification — terminal state is FAILED
        ctx.state = AgentState.FAILED

    # ── MEMORY ───────────────────────────────────────────────────────
    async def memory_update(self, ctx: AgentRunContext, req: Any, current_user: str) -> None:
        """Background: Memory Updater role extracts learnings from a terminal run.

        Terminal-state learning policy (review §4.2 L6): DONE runs record what
        worked; FAILED / NEEDS_REVIEW runs record what went wrong — failure is
        exactly the experience worth remembering. The run status stored with
        the experience is the *actual* terminal state, never a blanket "ok".
        """
        d = self.deps
        terminal = ctx.state.value if ctx.state in AGENT_TERMINAL_STATES else "UNKNOWN"
        outcome_hint = (
            "The task completed successfully."
            if ctx.state == AgentState.DONE
            else (
                f"The task ended as {terminal} — extract what went wrong and "
                "what to do differently next time, not a success story."
            )
        )
        context = (
            f"{d.memory_updater_prompt}\n\n"
            f"Task: {req.message}\n"
            f"Terminal status: {terminal}. {outcome_hint}\n\n"
            f"Last 5 transcript steps:\n{json.dumps(ctx.transcript[-5:], ensure_ascii=False)}"
        )
        try:
            raw = await d.generate(
                message="Extract learnings from this completed task.",
                context=context, max_tokens=self.phase_budgets.memory_tokens, temperature=0.1,
            )
            mem = extract_action(str(raw))
            kept_learnings = filter_learnings(mem.get("learnings") or [])
            if mem.get("save_to_knowledge") and kept_learnings:
                learnings = "\n".join(kept_learnings)
                status_label = {
                    AgentState.DONE: "ok",
                    AgentState.NEEDS_REVIEW: "needs_review",
                    AgentState.FAILED: "failed",
                }.get(ctx.state, "unknown")
                if d.brain_memory is not None:
                    # This runtime is LLM-driven — its learnings are real
                    # experiences and enter the brain with provenance.
                    d.brain_memory.record_experience(
                        f"Agent: {req.message[:60]}",
                        learnings,
                        run={
                            "mode": "llm",
                            "status": status_label,
                            "agent_id": "agent:executor",
                            "steps": len(ctx.transcript),
                        },
                        user_email=current_user or None,
                    )
                else:
                    d.knowledge_save(
                        learnings,
                        folder="30_Projects",
                        title=f"Agent: {req.message[:60]}",
                    )
        except Exception as exc:
            # Never crash a completed run, but never swallow silently either.
            logging.warning("agent memory update failed: %s", exc)

File: rust/lattice-agent/src/agentloop/recovery.rs
//! ROLLBACK — putting the workspace back, honestly.
//!
//! A port of `latticeai.core.agent.recovery`. Three tiers, tried in order: git
//! where the policy says the tool is git-rollbackable, then the pre-write
//! snapshot the executor captured, and otherwise `mode="none"` — an admission,
//! not a claim. Rollback always ends the run as FAILED, because it *is* the
//! recovery from a failed verification.
//!
//! This is the one phase that touches the workspace directly instead of going
//! through the worker seam. That is deliberate: a rollback that has to ask a
//! possibly-broken worker for permission to undo the worker's own writes is not
//! a rollback.

use std::time::Duration;

use serde_json::{json, Value};

use super::{RunRequest, Runtime, GIT_ROLLBACK_TIMEOUT_SECS};
use crate::pystr::{is_truthy, py_list_repr, py_str};
use crate::state::{AgentRunContext, AgentState};

impl Runtime {
    fn snapshot_for(ctx: &AgentRunContext, path: &str) -> Option<Value> {
        ctx.rollback_log
            .iter()
            .find(|entry| entry.get("path").and_then(Value::as_str) == Some(path))
            .cloned()
    }

    /// Recover one path: git when governed and available, else the pre-write
    /// snapshot, else an honest `mode="none"`.
    async fn rollback_one(&self, ctx: &AgentRunContext, path: &str, governance: &Value) -> Value {
        if governance.get("rollback").and_then(Value::as_str) == Some("git") {
            let mut result = self.git_checkout(path).await;
            if result.get("ok") == Some(&json!(true)) {
                result["mode"] = json!("git");
                return result;
            }
        }
        let snapshot = Self::snapshot_for(ctx, path);
        if let Some(snapshot) = snapshot {
            if !snapshot.get("too_large").is_some_and(is_truthy) {
                let existed = snapshot.get("existed").is_some_and(is_truthy);
                let content = if existed {
                    snapshot.get("content").and_then(Value::as_str)
                } else {
                    None
                };
                let mut restored = self.restore_snapshot(path, content);
                if restored.get("path").is_none() {
                    restored["path"] = json!(path);
                }
                restored["mode"] = json!("snapshot");
                return restored;
            }
        }
        json!({
            "path": path, "ok": false, "mode": "none",
            "error": "no rollback available (git not applicable, no usable snapshot)",
        })
    }

    /// `ToolDispatchService.rollback_file`: `git checkout -- <path>` under the
    /// agent root, with the same ten-second ceiling.
    async fn git_checkout(&self, path: &str) -> Value {
        let mut command = tokio::process::Command::new("git");
        command
            .args(["checkout", "--", path])
            .current_dir(self.deps.workspace.root())
            .stdin(std::process::Stdio::null());
        let spawned = tokio::time::timeout(
            Duration::from_secs(GIT_ROLLBACK_TIMEOUT_SECS),
            command.output(),
        )
        .await;
        match spawned {
            Ok(Ok(output)) => json!({
                "path": path,
                "ok": output.status.success(),
                "stderr": crate::pystr::char_slice(
                    &String::from_utf8_lossy(&output.stderr), 200,
                ),
            }),
            // Python wraps the call in `except Exception`, so a missing git or a
            // timeout is a failed recovery attempt, never a failed run.
            Ok(Err(error)) => json!({"path": path, "ok": false, "error": error.to_string()}),
            Err(_) => json!({
                "path": path, "ok": false,
                "error": format!("git checkout timed out after {GIT_ROLLBACK_TIMEOUT_SECS} seconds"),
            }),
        }
    }

    /// ROLLBACK: recover written files (git → snapshot → none), then FAILED.
    pub async fn rollback(&mut self, ctx: &mut AgentRunContext, req: &RunRequest) {
        let mut rolled: Vec<Value> = Vec::new();
        let mut seen: Vec<String> = Vec::new();
        for step in ctx.transcript.clone() {
            if step.get("state").and_then(Value::as_str) != Some(AgentState::Executing.as_str()) {
                continue;
            }
            let Some(result) = step.get("result").filter(|value| value.is_object()) else {
                continue;
            };
            let governance = step
                .get("governance")
                .filter(|value| is_truthy(value))
                .cloned()
                .unwrap_or_else(|| json!({}));
            let path = result
                .get("path")
                .filter(|value| is_truthy(value))
                .or_else(|| step.get("args").and_then(|args| args.get("path")))
                .filter(|value| is_truthy(value))
                .map(py_str)
                .unwrap_or_default();
            if path.is_empty() || seen.contains(&path) {
                continue;
            }
            let git_governed = governance.get("rollback").and_then(Value::as_str) == Some("git");
            let file_create = step
                .get("action")
                .and_then(Value::as_str)
                .is_some_and(|action| self.deps.file_create_actions.contains(action));
            if !git_governed && !file_create {
                continue;
            }
            seen.push(path.clone());
            rolled.push(self.rollback_one(ctx, &path, &governance).await);
        }

        ctx.transcript.push(json!({
            "state": AgentState::Rollback.as_str(),
            "rolled_back": rolled,
        }));
        let recovered_count = rolled
            .iter()
            .filter(|entry| entry.get("ok").is_some_and(is_truthy))
            .count();
        ctx.trace.decision(
            "rollback",
            "rolled_back",
            &[
                ("attempted", json!(rolled.len())),
                ("recovered", json!(recovered_count)),
            ],
        );
        let recovered: Vec<String> = rolled
            .iter()
            .filter(|entry| entry.get("ok").is_some_and(is_truthy))
            .map(|entry| {
                format!(
                    "{} ({})",
                    py_str(entry.get("path").unwrap_or(&Value::Null)),
                    py_str(entry.get("mode").unwrap_or(&Value::Null)),
                )
            })
            .collect();
        ctx.final_message = if recovered.is_empty() {
            "롤백을 시도했으나 복구할 파일이 없거나 git/스냅샷 복구 수단이 없습니다.".into()
        } else {
            // Python interpolates the *list*, so the message carries a Python
            // list literal — single quotes and all.
            format!(
                "실행 실패로 롤백했습니다. 복구 파일: {}",
                py_list_repr(&recovered)
            )
        };
        self.audit(
            "agent_rollback",
            &[
                ("user_email", json!(req.user_email)),
                ("rolled_back", json!(rolled)),
            ],
        );
        self.emit_step(
            "rollback",
            "rolled_back",
            &[("recovered", json!(recovered.len()))],
        );
        // Rollback is recovery from a failed verification — terminal is FAILED.
        ctx.state = AgentState::Failed;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::agentloop::harness::harness;
    use crate::agentloop::SNAPSHOT_MAX_BYTES;

    fn wrote(path: &str, rollback: &str) -> Value {
        json!({"state": "EXECUTING", "action": "write_file",
               "args": {"path": path}, "governance": {"rollback": rollback},
               "result": {"path": path, "bytes": 3}})
    }

    #[tokio::test]
    async fn a_snapshot_restores_prior_content_and_deletes_a_new_file() {
        let mut harness = harness(&[]).await;
        std::fs::write(harness.root.join("existing.md"), b"before").expect("seed");
        std::fs::write(harness.root.join("existing.md"), b"after").expect("overwrite");
        std::fs::write(harness.root.join("fresh.md"), b"new").expect("create");
        let mut ctx = harness.context();
        ctx.rollback_log = vec![
            json!({"path": "existing.md", "existed": true, "content": "before",
                   "too_large": false}),
            json!({"path": "fresh.md", "existed": false, "content": null, "too_large": false}),
        ];
        ctx.transcript = vec![wrote("existing.md", "none"), wrote("fresh.md", "none")];

        harness.runtime.rollback(&mut ctx, &harness.request).await;

        assert_eq!(ctx.state, AgentState::Failed, "rollback always ends FAILED");
        assert_eq!(
            std::fs::read_to_string(harness.root.join("existing.md")).expect("read"),
            "before"
        );
        assert!(
            !harness.root.join("fresh.md").exists(),
            "a created file is removed"
        );
        let rolled = ctx.transcript.last().expect("step")["rolled_back"].clone();
        assert_eq!(
            rolled[0],
            json!({"path": "existing.md", "ok": true,
                                     "action": "restored", "mode": "snapshot"})
        );
        assert_eq!(
            rolled[1],
            json!({"path": "fresh.md", "ok": true,
                                     "action": "deleted", "mode": "snapshot"})
        );
        assert_eq!(
            ctx.final_message,
            "실행 실패로 롤백했습니다. 복구 파일: ['existing.md (snapshot)', 'fresh.md (snapshot)']",
            "the message interpolates a Python list repr"
        );
        assert_eq!(harness.runtime.audit[0]["event"], "agent_rollback");
    }

    #[tokio::test]
    async fn without_a_usable_snapshot_the_answer_is_mode_none() {
        let mut harness = harness(&[]).await;
        let mut ctx = harness.context();
        ctx.rollback_log = vec![json!({"path": "big.md", "existed": true, "content": null,
                                       "too_large": true})];
        ctx.transcript = vec![wrote("big.md", "none"), wrote("unknown.md", "none")];
        harness.runtime.rollback(&mut ctx, &harness.request).await;
        let rolled = ctx.transcript.last().expect("step")["rolled_back"].clone();
        for entry in rolled.as_array().expect("rows") {
            assert_eq!(entry["mode"], "none");
            assert_eq!(entry["ok"], false);
            assert_eq!(
                entry["error"],
                "no rollback available (git not applicable, no usable snapshot)"
            );
        }
        assert_eq!(
            ctx.final_message,
            "롤백을 시도했으나 복구할 파일이 없거나 git/스냅샷 복구 수단이 없습니다."
        );
    }

    #[tokio::test]
    async fn only_git_governed_or_file_creating_steps_are_recovered() {
        let mut harness = harness(&[]).await;
        let mut ctx = harness.context();
        ctx.transcript = vec![
            // A read with a result and a path: not a file create, not git.
            json!({"state": "EXECUTING", "action": "read_file", "args": {"path": "skip.md"},
                   "governance": {"rollback": "none"}, "result": {"path": "skip.md"}}),
            // A blocked step has no result at all.
            json!({"state": "EXECUTING", "action": "write_file", "args": {"path": "blocked.md"},
                   "error": "BLOCKED"}),
            wrote("kept.md", "none"),
            // The same path twice is recovered once.
            wrote("kept.md", "none"),
        ];
        harness.runtime.rollback(&mut ctx, &harness.request).await;
        let rolled = ctx.transcript.last().expect("step")["rolled_back"]
            .as_array()
            .expect("rows")
            .clone();
        assert_eq!(rolled.len(), 1);
        assert_eq!(rolled[0]["path"], "kept.md");
    }

    #[tokio::test]
    async fn a_git_governed_path_falls_back_to_the_snapshot_when_git_cannot_help() {
        // The workspace is not a git repository, so `git checkout` fails and the
        // snapshot tier answers — which is exactly the honest ordering.
        let mut harness = harness(&[]).await;
        std::fs::write(harness.root.join("tracked.md"), b"changed").expect("write");
        let mut ctx = harness.context();
        ctx.rollback_log = vec![json!({"path": "tracked.md", "existed": true,
                                       "content": "original", "too_large": false})];
        ctx.transcript = vec![wrote("tracked.md", "git")];
        harness.runtime.rollback(&mut ctx, &harness.request).await;
        let rolled = ctx.transcript.last().expect("step")["rolled_back"].clone();
        assert_eq!(rolled[0]["mode"], "snapshot");
        assert_eq!(
            std::fs::read_to_string(harness.root.join("tracked.md")).expect("read"),
            "original"
        );
    }

    #[tokio::test]
    async fn nothing_to_roll_back_is_still_a_recorded_rollback() {
        let mut harness = harness(&[]).await;
        let mut ctx = harness.context();
        harness.runtime.rollback(&mut ctx, &harness.request).await;
        assert_eq!(
            ctx.transcript,
            vec![json!({"state": "ROLLBACK", "rolled_back": []})]
        );
        assert_eq!(ctx.state, AgentState::Failed);
    }

    #[test]
    fn a_snapshot_of_an_escaping_path_says_so_instead_of_reading_it() {
        let dir = tempfile::tempdir().expect("tempdir");
        let workspace =
            crate::sandbox::Workspace::new(dir.path().join("agent_workspace")).expect("ws");
        std::fs::write(dir.path().join("outside.txt"), b"secret").expect("outside");
        let runtime = Runtime::new(crate::agentloop::LoopDeps::new(
            crate::worker::WorkerClient::new("http://127.0.0.1:1"),
            workspace,
        ));
        assert_eq!(
            runtime.snapshot_file("../outside.txt"),
            json!({"existed": false, "content": null, "too_large": false,
                   "error": "path escapes the agent workspace"})
        );
        assert_eq!(
            runtime.restore_snapshot("../outside.txt", Some("x")),
            json!({"path": "../outside.txt", "ok": false,
                   "error": "path escapes the agent workspace"})
        );
        assert_eq!(
            std::fs::read_to_string(dir.path().join("outside.txt")).expect("read"),
            "secret"
        );
    }

    #[test]
    fn a_snapshot_over_the_cap_records_the_size_not_the_bytes() {
        let dir = tempfile::tempdir().expect("tempdir");
        let workspace =
            crate::sandbox::Workspace::new(dir.path().join("agent_workspace")).expect("ws");
        let big = workspace.root().join("big.bin");
        std::fs::write(&big, vec![b'a'; (SNAPSHOT_MAX_BYTES + 1) as usize]).expect("big");
        std::fs::create_dir_all(workspace.root().join("adir")).expect("dir");
        let runtime = Runtime::new(crate::agentloop::LoopDeps::new(
            crate::worker::WorkerClient::new("http://127.0.0.1:1"),
            workspace,
        ));
        assert_eq!(
            runtime.snapshot_file("big.bin"),
            json!({"existed": true, "content": null, "too_large": true})
        );
        assert_eq!(
            runtime.snapshot_file("adir"),
            json!({"existed": false, "content": null, "too_large": false}),
            "a directory is not a file to snapshot"
        );
        assert_eq!(
            runtime.snapshot_file("missing.md"),
            json!({"existed": false, "content": null, "too_large": false})
        );
    }
}
